Description: Download the PGO training file earlier in the build sequence, because Launchpad builders have a timeout for how long they are allowed to access the internet (through a proxy) starting from the start of the build.
Author: Olivier Tilloy <olivier.tilloy@canonical.com>

--- a/tools/clang/scripts/build.py
+++ b/tools/clang/scripts/build.py
@@ -482,6 +482,15 @@ def CopyLibstdcpp(args, build_dir):
       CopyFile(libstdcpp, libpath)
 
 
+def DownloadPGOTrainingFile():
+  curdir = os.getcwd()
+  os.chdir(THIRD_PARTY_DIR)
+  training_source = 'pgo_training-1.ii'
+  with open(training_source, 'wb') as f:
+    DownloadUrl(CDS_URL + '/' + training_source, f)
+  os.chdir(curdir)
+
+
 def gn_arg(v):
   if v == 'True':
     return True
@@ -620,6 +629,9 @@ def main():
   AddCMakeToPath(args)
   DeleteChromeToolsShim()
 
+  if args.pgo:
+    DownloadPGOTrainingFile()
+
 
   if args.skip_build:
     return 0
@@ -888,8 +900,7 @@ def main():
     # from PGO as well. Perhaps the training could be done asynchronously by
     # dedicated buildbots that upload profiles to the cloud.
     training_source = 'pgo_training-1.ii'
-    with open(training_source, 'wb') as f:
-      DownloadUrl(CDS_URL + '/' + training_source, f)
+    shutil.copy(os.path.join(THIRD_PARTY_DIR, training_source), os.getcwd())
     train_cmd = [os.path.join(LLVM_INSTRUMENTED_DIR, 'bin', 'clang++'),
                 '-target', 'x86_64-unknown-unknown', '-O2', '-g', '-std=c++14',
                  '-fno-exceptions', '-fno-rtti', '-w', '-c', training_source]
